File: Emoticon/Model/EmoticonPackage.m
//
//  EmoticonPackage.m
//  表情键盘OC
//
//  Created by 单车 on 2019/10/29.
//  Copyright © 2019 单车. All rights reserved.
//

#import "EmoticonPackage.h"

@implementation EmoticonPackage

/** _emoticons */
- (NSMutableArray<Emoticon *> *)emoticons{
    if (_emoticons == nil) {
        _emoticons = [NSMutableArray array];
    }
    return _emoticons;
}

- (instancetype)initWithId:(NSString *)idString{
    self = [super init];
    if (self) {
        //最近分组
        if ([idString isEqualToString:@""]) {
            [self addEmptyEmotion:YES];
            return self;
        }
        
        //2.根据id 拼接info.plist的路径
        NSString *plistPath = [[NSBundle mainBundle]pathForResource:[NSString stringWithFormat:@"%@/info.plist",idString] ofType:nil inDirectory:@"Emoticons.bundle"];
        
        NSDictionary *dict = [NSDictionary dictionaryWithContentsOfFile:plistPath];
        
        NSArray *array = dict[@"emoticons"];
        
        NSInteger index = 0;
        for (NSMutableDictionary *dict in array) {
            NSString *png = dict[@"png"];
            if (png != nil) {
                [dict setValue:[NSString stringWithFormat:@"%@/%@",idString,png] forKey:@"png"];
            }
            [self.emoticons addObject:[[Emoticon alloc]initWithDict:dict]];
            index += 1;
            
            if (index == 20) {
                [self.emoticons addObject:[[Emoticon alloc]initWithRemove:YES]];
                
                //添加删除表情
                index = 0;
            }
        }
        [self addEmptyEmotion:NO];
    }
    return self;
}


/// 添加空的表情
/// @param isRecently 是否为最近
- (void)addEmptyEmotion:(BOOL)isRecently{
    NSInteger count = self.emoticons.count % 21;
    if (count == 0 && isRecently == NO) {
        return;
    }
    for (NSInteger i = count; i < 20; i ++) {
        [self.emoticons addObject:[[Emoticon alloc]initWithEmpty:YES]];
    }
    [self.emoticons addObject:[[Emoticon alloc]initWithRemove:YES]];
}
@end

File: 09-表情键盘/09-表情键盘/Emoticon/Model/EmoticonPackage.swift
//
//  EmoticonPackage.swift
//  09-表情键盘
//
//  Created by 单车 on 2019/10/22.
//  Copyright © 2019 单车. All rights reserved.
//

import UIKit

class EmoticonPackage: NSObject {
    var emoticons : [Emoticon] = [Emoticon]()
    
    init(id : String) {
        super.init()
        //1.最近分组
        if id == ""{
            addEmptyEmotion(isRecently: true)
            return
        }
        //2.根据id拼接info.plist的路径
        let plistPath = Bundle.main.path(forResource: "\(id)/info.plist", ofType: nil, inDirectory: "Emoticons.bundle")!
        
        let dict = NSDictionary(contentsOfFile: plistPath)!

        //3.根据plist文件路径读取数据 [[String:String]]
//        let array = NSArray(contentsOfFile: plistPath)! as! [[String:String]]
        
        let array = dict["emoticons"] as! [[String:String]]

        //4.遍历数组
        var index = 0
        
        for var dict in array {
            if let png = dict["png"]{
                dict["png"] = id + "/" + png
            }
            emoticons.append(Emoticon(dict: dict))
            index += 1
            
            if index == 20 {
                emoticons.append(Emoticon(isRemove: true))
                
                //添加删除表情
                index = 0
            }
        }
        
        //5.添加空白表情
        addEmptyEmotion(isRecently: false)
    }
    private func addEmptyEmotion(isRecently:Bool){
        let count = emoticons.count % 21
        if count == 0 && !isRecently{
            return
        }
        for _ in count..<20 {
            emoticons.append(Emoticon(isEmpty: true))
        }
        emoticons.append(Emoticon(isRemove: true))
        
    }
}
